### PR TITLE
Use single function to calculate changeset state/review state

### LINKED ISSUE #9101: Use the same code for burndown chart and derived Changeset State
Right now `CalcCounts` and, for example, `ChangsetEvents.ReviewState` use different code, which lead to a lot of bugs sneaking in in the past.

We need to find a solution to merge both pieces of critical business logic.

### PR DESCRIPTION
**Pull request by mrnugget**
This fixes #9101 by introducing a single function, used by `CalcCounts` (the burndown chart) and `ComputeReviewState`/`ComputeState` (used to calculate the state/review-state of a single changeset).

As described in #9101 we previously ran into a lot of errors due to these two parts of the codebase — burndown chart and the calculation of state for a single changeset — using different ways to derive state from a changeset and its events.

Now we have a single function for this.

### CODE REVIEW

File: enterprise/internal/campaigns/changeset_history.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+		})
+	}
+
+	openedAt := ch.ExternalCreatedAt()
+	if openedAt.IsZero() {
+		return nil, errors.New("changeset ExternalCreatedAt has zero value")
+	}
+	pushStates(openedAt)
+
+	for _, e := range ce {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/changeset_history.go:79]: Everything in this loop here has just been moved from the previous `CalcCounts` function and slightly adjusted to use `pushStates` etc.

File: enterprise/internal/campaigns/changeset_history.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+// computeHistory calcuates the changesetHistory for the given Changeset and
+// its ChangesetEvents.
+// The ChangesetEvents MUST be sorted by their Timestamp.
+func computeHistory(ch *cmpgn.Changeset, ce ChangesetEvents) (changesetHistory, error) {
+	if len(ce) > 1 {
+		first, last := ce[0], ce[len(ce)-1]
+		if first.Timestamp().After(last.Timestamp()) {
+			return nil, errors.New("changeset events no ordered by timestamps")
```
Comment (ryanslade) [inline enterprise/internal/campaigns/changeset_history.go:52]:
```suggestion
			return nil, errors.New("changeset events not ordered by timestamps")
```

Comment (ryanslade) [reply enterprise/internal/campaigns/changeset_history.go:52]: Also, the last item might be after the first one but the ones between could still be in the wrong order. Maybe you can use `sort.IsSorted`?

**Comment by ryanslade** [comment]
Nice, this is going to be useful as I need to add a new case for https://github.com/sourcegraph/sourcegraph/issues/9139

File: enterprise/internal/campaigns/changeset_history.go (hunk 2 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/changeset_history.go:52]: TIL about `sort.IsSorted`!

**Pull request merged**